### PR TITLE
extra is now an object {re: string, optional: boolean}

### LINKED ISSUE #39: [Enhancement] Consider removing the appended wildcard [ ]?(.*)
Although it is well documented, having HE append `[ ]?(.*)` to the base string auto-magically is very confusing IMO.

With the non-negotiable wildcard at the end, we're exposing devs to all kinds of subtle errors.

For example, suppose we have two actions.
1) "list something"
2) "list something in another format"

Action 1 and action 2 are translated by HE to /list something[ ]?(.*)?/ and /list something in another format[ ]?(.*)?/ respectively. Now, if I send the message "list something in another format" won't both listeners be triggered? This seems to happen on my end.

### LINKED ISSUE #40: [Feature] Optional extra parameter
As we previously discussed, we need better support for actions defined with regular expressions.

With the current structure of `robot.e.create`, it's not clear how a message with multiple regex criteria, such as `list defects discovered by (.+) in sprint ([^ ]+)(?: with filter(?:s)? (.*))?`, would fit within the structure. Furthermore, for my use case, adding a whitespace between the `action` and the `extra`, does not allow me to make the `extra` param optional.

### PR DESCRIPTION
**Pull request by mosheBelostotsky**
extra is now an object, full BW compatibility retained
Close #40, Close #39

### CODE REVIEW

**Comment by rqc** [comment]
@mosheBelostotsky can we rename `extra` to `regex_suffix`. I think it is more self-explanatory. What do you think?

File: src/0_bootstrap.coffee [hunk 1]
```diff
@@ -74,6 +74,30 @@ module.exports = (robot) ->
...
+  #  extra: extra element
+  #    optional: true/false- should it be optional
+  #    re: string that representing the regex (optional)
+  build_extra_re = (info) ->
+    extra = info.extra
+    # TODO: prevent calls similarity as much as possible
+    # TODO: only one varb+entity may have optional: true
+    # TODO: forbid {optional: true} with {re: null, optional: false}
+    # TODO: try to check that 2 regexps are not equal (at least no the same)
+    if extra.re
```
Comment (rqc) [inline src/0_bootstrap.coffee]: verify `typeof` `re`

File: test/enterprise-test.coffee [hunk 2]
```diff
@@ -154,3 +156,114 @@ describe 'enterprise tests', ->
...
+    it 'extra, no regex, optional: false- call without extra', ->
+      call = extend({extra: {optional: false}}, common_create)
+      @room.robot.e.create call, callback
+      @room.user.say('alice', '@hubot foo verb entity').then =>
+        expect(@room.messages).to.eql [
+          [ 'alice', '@hubot foo verb entity with me' ],
+          [ 'hubot', '@alice NONE' ]
+        ]
+
+  it 'no extra boject- call with extra', ->
```
Comment (rqc) [inline test/enterprise-test.coffee]: typo

File: test/enterprise-test.coffee [hunk 3]
```diff
@@ -154,3 +156,114 @@ describe 'enterprise tests', ->
...
+    @room.robot.e.create common_create, callback
+    @room.user.say('alice', '@hubot foo verb entity with me').then =>
+      expect(@room.messages).to.eql [
+        [ 'alice', '@hubot foo verb entity with me' ],
+        [ 'hubot', '@alice with me' ]
+      ]
+
+  it 'no extra boject- call without extra', ->
```
Comment (rqc) [inline test/enterprise-test.coffee]: - [ ] typo

File: test/enterprise-test.coffee [hunk 4]
```diff
@@ -154,3 +156,114 @@ describe 'enterprise tests', ->
...
+        [ 'hubot', '@alice me' ]
+      ]
+
+  it 'extra passed as string (backward compatibility)- call without extra', ->
+    common_create.extra = "with (.*)"
+    @room.robot.e.create common_create, callback
+    @room.user.say('alice', '@hubot foo verb entity').then =>
+      expect(@room.messages).to.eql [
+        [ 'alice', '@hubot foo verb entity' ]
+      ]
```
Comment (rqc) [inline test/enterprise-test.coffee]: - [ ] add tests for bounds (e.g. expected prefixed `space`)

File: src/0_bootstrap.coffee [hunk 5]
```diff
@@ -74,6 +74,30 @@ module.exports = (robot) ->
...
+  # build extra part of the regex
+  #
+  # info: info object from build_enterprise_regex
+  #  extra: extra element
+  #    optional: true/false- should it be optional
+  #    re: string that representing the regex (optional)
+  build_extra_re = (info) ->
+    extra = info.extra
+    # TODO: prevent calls similarity as much as possible
+    # TODO: only one varb+entity may have optional: true
```
Comment (rqc) [inline src/0_bootstrap.coffee]: - [ ] typo _verb_

File: src/0_bootstrap.coffee [hunk 6]
```diff
@@ -74,6 +74,30 @@ module.exports = (robot) ->
...
+    # TODO: forbid {optional: true} with {re: null, optional: false}
+    # TODO: try to check that 2 regexps are not equal (at least no the same)
+    if extra.re
+      # if extra.re passed and its optional
+      if extra.optional
+        return "(?: #{extra.re})?"
+      #if it's not optional
+      return " #{extra.re}"
+    #if no extra.re and optional
```
Comment (rqc) [inline src/0_bootstrap.coffee]: - [ ] can you add an `else` for readability?

File: src/0_bootstrap.coffee (hunk 1 above)
Comment (rqc) [inline src/0_bootstrap.coffee]: what would happen here if `extra` is `undefined`?

File: test/enterprise-test.coffee [hunk 7]
```diff
@@ -154,3 +156,114 @@ describe 'enterprise tests', ->
...
+    @room = helper.createRoom()
+
+  # common create object, to extend
+  common_create = {product: 'foo', verb: 'verb', entity: 'entity'}
+
+  afterEach ->
+    @room.destroy()
+
+  it 'extra: regex, optional: true- call with extra', ->
+    call = extend({extra: {re: 'with (.*)', optional: true}}, common_create)
```
Comment (rqc) [inline test/enterprise-test.coffee]: - [ ] please add variations in which `extra` is `undefined`, `nil`, or some other `type`. Try different variations of the object structure.

**Comment by rqc** [comment]
@mosheBelostotsky looking good... please read the comments and make the appropriate changes. Thanks! 👍

File: src/0_bootstrap.coffee (hunk 1 above)
Comment (mosheBelostotsky) [inline src/0_bootstrap.coffee]: cannot be. verified in the previous function.

**Comment by mosheBelostotsky** [comment]
@rqc I prefer to preserve `extra` for BW compitability...

**Comment by rqc** [comment]
@mosheBelostotsky we can still support `extra`, but start migrating to using something more readable like `regex_suffix`. That way it will be backwards compatible and forward readable 😉

**Comment by mosheBelostotsky** [comment]
@rqc WILCO :wink:

File: src/0_bootstrap.coffee (hunk 1 above)
Comment (mosheBelostotsky) [inline src/0_bootstrap.coffee]: Moving all the checks extra from `robot.e.create` to `build_extra_re`

File: test/enterprise-test.coffee (hunk 4 above)
Comment (mosheBelostotsky) [inline test/enterprise-test.coffee]: Added tests without `space`

File: src/0_bootstrap.coffee (hunk 5 above)
Comment (mosheBelostotsky) [inline src/0_bootstrap.coffee]: done

File: src/0_bootstrap.coffee (hunk 6 above)
Comment (mosheBelostotsky) [inline src/0_bootstrap.coffee]: done

File: test/enterprise-test.coffee (hunk 2 above)
Comment (mosheBelostotsky) [inline test/enterprise-test.coffee]: done

File: test/enterprise-test.coffee (hunk 3 above)
Comment (mosheBelostotsky) [inline test/enterprise-test.coffee]: done

File: test/enterprise-test.coffee (hunk 7 above)
Comment (mosheBelostotsky) [inline test/enterprise-test.coffee]: Done

File: test/enterprise-test.coffee [hunk 8]
```diff
@@ -174,94 +174,159 @@ describe 'listener extra object tests', ->
...
+    delete common_create.regex_suffix
     @room.robot.e.create common_create, callback
     @room.user.say('alice', '@hubot foo verb entity with me').then =>
       expect(@room.messages).to.eql [
         [ 'alice', '@hubot foo verb entity with me' ],
         [ 'hubot', '@alice me' ]
       ]
 
-  it 'extra passed as string (backward compatibility)- call without extra', ->
+  it 'extra passed as string (backward compatibility)- call without suffix', ->
```
Comment (rqc) [inline test/enterprise-test.coffee]: Good call!! :+1:

**Comment by rqc** [comment]
Looks great. I am merging.

**Pull request merged**